pgx/src/spi.rs

```diff
@@ -160,6 +160,162 @@ impl SpiConnection {
+        limit: Option<i64>,
+        arguments: Self::Arguments,
+    ) -> Self::Result;
+
+    /// Open a cursor for the query
+    fn open_cursor<'c: 'cc, 'cc>(
+        self,
+        client: &'cc SpiClient<'c>,
+        args: Self::Arguments,
+    ) -> Result<SpiCursor<'c>, Error>;
+}
+
+impl<'a> Query for &'a String {
+    type Arguments = Option<Vec<(PgOid, Option<pg_sys::Datum>)>>;
+    type Result = SpiTupleTable;
+
+    fn execute(
+        self,
+        client: &SpiClient,
+        read_only: bool,
+        limit: Option<i64>,
+        arguments: Self::Arguments,
+    ) -> Self::Result {
+        self.as_str().execute(client, read_only, limit, arguments)
+    }
+
+    fn open_cursor<'c: 'cc, 'cc>(
+        self,
+        client: &'cc SpiClient<'c>,
+        args: Self::Arguments,
+    ) -> Result<SpiCursor<'c>, Error> {
+        self.as_str().open_cursor(client, args)
+    }
+}
+
+impl<'a> Query for &'a str {
+    type Arguments = Option<Vec<(PgOid, Option<pg_sys::Datum>)>>;
+    type Result = SpiTupleTable;
+
+    fn execute(
+        self,
+        _client: &SpiClient,
+        read_only: bool,
+        limit: Option<i64>,
+        arguments: Self::Arguments,
+    ) -> Self::Result {
+        unsafe {
+            pg_sys::SPI_tuptable = std::ptr::null_mut();
+        }
+
+        let src = std::ffi::CString::new(self).expect("query contained a null byte");
+        let status_code = match arguments {
+            Some(args) => {
+                let nargs = args.len();
+                let mut argtypes = vec![];
+                let mut datums = vec![];
+                let mut nulls = vec![];
+
+                for (argtype, datum) in args {
+                    argtypes.push(argtype.value());
+
+                    match datum {
+                        Some(datum) => {
+                            // ' ' here means that the datum is not null
+                            datums.push(datum);
+                            nulls.push(' ' as std::os::raw::c_char);
+                        }
+
+                        None => {
+                            // 'n' here means that the datum is null
+                            datums.push(pg_sys::Datum::from(0usize));
+                            nulls.push('n' as std::os::raw::c_char);
+                        }
+                    }
+                }
+
+                unsafe {
+                    pg_sys::SPI_execute_with_args(
+                        src.as_ptr(),
+                        nargs as i32,
+                        argtypes.as_mut_ptr(),
+                        datums.as_mut_ptr(),
+                        nulls.as_ptr(),
+                        read_only,
+                        limit.unwrap_or(0),
+                    )
+                }
+            }
+            None => unsafe { pg_sys::SPI_execute(src.as_ptr(), read_only, limit.unwrap_or(0)) },
+        };
+
+        SpiClient::prepare_tuple_table(status_code)
+    }
+
+    fn open_cursor<'c: 'cc, 'cc>(
```

Comment:
This function and `execute()` are essentially the same.  Need to find a way to eliminate all the code duplication.  And I suspect there's more...

Reply:
They are very similar but not the same. There are subtle differences. At the very minimum, we should extract this:

```rust
   for (argtype, datum) in args {
                    argtypes.push(argtype.value());

                    match datum {
                        Some(datum) => {
                            // ' ' here means that the datum is not null
                            datums.push(datum);
                            nulls.push(' ' as std::os::raw::c_char);
                        }

                        None => {
                            // 'n' here means that the datum is null
                            datums.push(pg_sys::Datum::from(0usize));
                            nulls.push('n' as std::os::raw::c_char);
                        }
                    }
                }
```

However, before we do any of this, let's agree on the surface of the PR first. We can either work on code quality improvement in the same PR after that or in a follow-up PR.

Reply:
Please consolidate as much of the duplicated code as you can in this PR.

Reply:
Does this mean you agree with the API's shape?

Reply:
Working on the consolidation patch.

Reply:
I've extracted the most gnarly part of the repetitiveness. Because of subtle dissimilarities, there's still some similarly-looking code, but I don't see a beautiful way to make abstract it further, at least not seeing this worth the time at the moment.